Gems/Terrain/Assets/Shaders/Terrain/TerrainDetailHelpers.azsli

```diff
@@ -257,46 +257,42 @@ void BlendDetailMaterial(inout DetailSurface surface, uint material1, uint mater
 {
     DetailSurface tempSurface;
     GetDetailSurfaceForMaterial(tempSurface, material1, detailUv);
+    WeightDetailSurface(tempSurface, weight * (1.0 - blend));
+    AddDetailSurface(surface, tempSurface);
     if (material2 != 0xFF)
     {
-        WeightDetailSurface(tempSurface, weight * blend);
-        AddDetailSurface(surface, tempSurface);
-
         GetDetailSurfaceForMaterial(tempSurface, material2, detailUv);
-        WeightDetailSurface(tempSurface, weight * (1.0 - blend));
-        AddDetailSurface(surface, tempSurface);
-    }
-    else
-    {
-        WeightDetailSurface(tempSurface, weight);
+        WeightDetailSurface(tempSurface, weight * blend);
         AddDetailSurface(surface, tempSurface);
     }
 }
 
 /*
 Populates a DetailSurface with material data gathered form the 4 nearest samples to detailMaterialIdUv. The weight
 of each detail material's contribution is calculated based on the distance to the center point for that sample (for
-instance, if detailMaterialIdUv falls perfectly inbetween all 4 samples, then each sample will be weighed at 25%).
+instance, if detailMaterialIdUv falls perfectly in-between all 4 samples, then each sample will be weighed at 25%).
 Each sample can have two different detail materials defined with a blend value to determine their relative contribution.
-The detailUv is used for samping the textures of each detail material.
+The detailUv is used for sampling the textures of each detail material.
 */
 bool GetDetailSurface(inout DetailSurface surface, float2 detailMaterialIdUv, float2 detailUv)
 {
     float2 textureSize;
     TerrainSrg::m_detailMaterialIdImage.GetDimensions(textureSize.x, textureSize.y);
 
-    float2 detailMaterialIdCoord = detailMaterialIdUv * textureSize + 0.5;
-    float2 detailMaterailIdUv = detailMaterialIdCoord;
-    float2 gatherWeight = frac(detailMaterialIdCoord);
+    float2 detailMaterialIdCoord = detailMaterialIdUv * textureSize; // uv -> pixel coordinate
```

Comment:
While I agree with what you're saying for regular textures, this case is a little different because the texture is generated on the CPU. The important thing is that the world position used to generate data for a specific pixel in the detail texture lines up with the world position being sampled in the shader. Since the texture is generated using integer coordinates it's better to think of the samples as being in the top-left corner of the pixel instead of the center. In other words, the 0,0 coordinate in the texture was generated from data gathered at the very corner of the texture bounds, not from a half-pixel offset from the corner as usual.

This means that the detail texture points should line up with integer coordinates in world space, so just to make sure I created a little sphere and checked at various integer coordinates to make sure it looked like it was positioned on the border between two material types and it looked like it was, so I'm fairly confident that it is correct. We will definitely build better test cases in the future to validate though - ultimately we need to create an automated test to make sure we're not accidentally sampling underlying data at an offset or rendering it at an offset etc, and also make sure a subtle bug like an offset doesn't crop up in the future.

Reply:
Thanks for clarifying. LGTM